sql/hive/src/main/scala/org/apache/spark/sql/hive/HiveExternalCatalog.scala

```diff
@@ -1175,6 +1205,27 @@ private[spark] class HiveExternalCatalog(conf: SparkConf, hadoopConf: Configurat
     client.listFunctions(db, pattern)
   }
 
+  /** Detect whether a table is stored with Hive-compatible metadata. */
+  private def isHiveCompatible(table: CatalogTable): Boolean = {
```

Comment:
There's already code to check compatibility in `createDataSourceTable`. The problem is that the logic there can only really be applied in the context of that method. Later on you cannot detect compatibility the same way, and just checking the schema is *not* enough.

The previous code was prone to leave corrupted tables (in old versions of Hive) or just plain not work (in newer versions of Hive). Having a consistent way of checking for Hive compatibility is key to that fix, because `alterTableSchema` has to behave differently for each case.